inc/VcfReader.hpp

```diff
@@ -0,0 +1,242 @@
+     * @param is_diploid TRUE iff there are two copies of `chromosome`;
+     * @param n_samples_in_vcf expected number of samples in every line;
+     * @param pass_only calls with FILTER different from PASS or `.` are discarded by the user;
+     * @param high_qual_only calls with `QUAL<min_qual` are discarded by the user;
+     * @param skip_samples sample GT fields are discarded by the user;
+     * @param min_sv_length calls shorter than this are discarded by the user;
+     * @param min_allele_frequency calls with too few haplotypes containing the ALT allele are discarded by the user;
+     * the procedure assumes that every sample has the same ploidy, so it might not work for CNVs or if a sample is a
+     * haploid reference;
+     * @param min_nonmissing_frequency calls with too few non-missing haplotypes are discarded by the user;
+     * the procedure assumes that every sample has the same ploidy, so it might not work for CNVs or if a sample is a
+     * haploid reference.
+     */
+    VcfRecord(const string& chromosome, bool is_diploid, bool high_qual_only, float min_qual, bool pass_only, uint32_t min_sv_length, bool skip_samples, uint32_t n_samples_in_vcf, float min_allele_frequency, float min_nonmissing_frequency);
+
+    /**
+     * Reads `stream` until EOL/EOF and loads some or all of the data in the current line.
+     *
+     * - If the call is a symbolic INS, no field after ALT is loaded.
+     * - If there are zero or more than one ALT, no field after ALT is loaded.
+     * - If `HIGH_QUAL_ONLY` is true and the call has low quality, no field after QUAL is loaded.
+     * - If `PASS_ONLY` is true and the call is not PASS, no field after FILTER is loaded.
+     * - If the call is not of a supported type, no field after INFO is loaded.
+     * - If the call is shorter than `MIN_SV_LENGTH`, no field after INFO is loaded.
+     * - If there are more samples than `N_SAMPLES_IN_VCF`, no sample after the maximum is loaded.
+     *
+     * @param tmp_* reused temporary space.
+     */
+    void set(ifstream& stream, string& tmp_buffer_1, string& tmp_buffer_2, pair<uint8_t, uint8_t>& tmp_pair);
+
+    /**
+     * @return TRUE iff the record passes all the constraints set at construction time.
+     */
+    bool passes_constraints() const;
+
+    /**
+     * Appends to `stream` a string representation of the object. No terminator is added at the end.
+     */
+    void print(ostream& stream) const;
+
+    /**
+     * Remark: the procedure handles the case where there are two copies of the chromosome but the GT field contains
+     * just one value (this can happen with CNVs or when the sample is a haploid reference).
+     *
+     * @param sample sample ID (the first sample has ID zero);
+     * @param out output pair that stores the only GT haplotype ID in `first`, or the two haplotype IDs in `first` and
+     * `second`, depending on the ploidy of `chrom`; -1 = missing haplotype;
+     * @param tmp_buffer reused temporary space;
```

Comment:
I'm not totally clear on how `tmp_buffer` is used, but if it is only for the internal operations of VcfRecord then it should probably be a private member of the class, instead of explicitly passed around by reference as an arg